taichi/analysis/value_diff.cpp

```diff
@@ -44,8 +44,19 @@ class ValueDiffLoopIndex : public IRVisitor {
   }
 
   void visit(LoopIndexStmt *stmt) override {
-    results[stmt->instance_id] =
-        DiffRange(stmt->loop == loop && stmt->index == loop_index, 1, 0);
+    results[stmt->instance_id] = DiffRange();
+    if (stmt->loop == loop && stmt->index == loop_index) {
+      results[stmt->instance_id] =
+          DiffRange(/*related=*/true, /*coeff=*/1, /*low=*/0);
+    } else if (auto range_for = stmt->loop->cast<RangeForStmt>()) {
+      if (range_for->begin->is<ConstStmt>() &&
+          range_for->end->is<ConstStmt>()) {
+        results[stmt->instance_id] = DiffRange(
+            /*related=*/true, /*coeff=*/0,
+            /*low=*/range_for->begin->as<ConstStmt>()->val[0].val_int(),
```

Comment:
BTW, this is incorrect if we iterate over a range in the decreasing order?

Reply:
Good catch! I'll fix it in another PR.

Reply:
Oh, even in decreasing order, we have `begin < end`... Nevertheless, I'll use min(begin, end - 1) to reduce confusion here.

Reply:
See also https://github.com/taichi-dev/taichi/pull/2276/files#diff-917d9436dcaafa0f1e41ae9bad90273a303f036f00da94e417788a7fa1dc5260R640-R643